### PR TITLE
feat(core): Add click to pay support in hyperswitch

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [x] New feature

## Description

Add click to pay and unified authentication service integration in core

## How did you test it?

check payments sanity

#1 : Create Business profile with `"is_click_to_pay_enabled": true`

```
curl --location 'http://localhost:8080/account/cewqcewc/business_profile' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
      "profile_name": "cewqcewc",
    
    "payment_link_config": {
        
        "business_specific_configs": {
            "key1": {
                "theme": "#3B845E",
                "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
                "seller_name": "sahkal",
                "display_sdk_only": true
            },
            "key2": {
                "theme": "#B202FF",
                "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
                "seller_name": "sahkal",
                "enable_saved_payment_method":true
                
            }

        }
    },
    "is_click_to_pay_enabled": true
}'
```

#2: Create Merchant connector account for mastercard click to pay and adyen (if authorisation supported by adyen only)

Ctp MCA
```
curl --location 'http://localhost:8080/account/cewqcewc/connectors' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '
{
    "connector_type": "authentication_processor",
    "connector_name": "ctp_mastercard",
    "connector_account_details": {
        "auth_type": "HeaderKey",
        "api_key": "MASTERCARD-API-KEY"
    },
    "test_mode": true,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "card_networks": [
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "debit",
                    "card_networks": [
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
    "metadata": {
        "city": "NY",
        "unit": "245",
        "live_endpoint_prefix": "vsfbjh"
    },
    "business_country": "US",
    "business_label": "default"
}
'
```
#3 : Do payments Create with `confirm=false`

```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_07eZpksxjV2V4SeM7X70PfZiQtsp27ygyhooXVjdV7mVIRxVKlELWwOrtVJ7XGRo' \
--data '{
    "amount": 1130,
    "currency": "USD",
    "confirm": false,
    "customer_id": "sahkal12",
    "return_url": "https://hyperswitch.io"
}'
```

#4. Get the client secret and do a confirm call

```
curl --location 'http://localhost:8080/payments/pay_VjKiaXwRYsaw8p5gLY2k/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_10d1a8db20d74c1b941685208ecd6a46' \
--data '
{
      "payment_method": "card",
    "payment_method_type": "debit",
    "client_secret": "pay_VjKiaXwRYsaw8p5gLY2k_secret_1QGBIB0svvM5xCMGA1my",
    "ctp_service_details" : {
        "merchant_transaction_id": "0a4e0d3.34f4a04b.803eb38b0a56f082296538d4a70c064a6b320639",
        "correlation_id": "34f4a04b.bcd1ac39-e0f3-479e-b3ee-703ade0c66dc",
        "x_src_flow_id" : "34f4a04b.bcd1ac39-e0f3-479e-b3ee-703ade0c66dc.1727438793"
    },
    "browser_info": {
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "color_depth": 24,
        "java_enabled":true,
        "java_script_enabled": true,
        "language": "en-GB",
        "screen_height":  1440,
        "screen_width": 2560,
        "time_zone": -330,
        "user_agent": "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Mobile Safari/537.36"
    }
}'
```

# Logs

PreAuthentication via Uas

Screenshot 2024-12-13 at 9 00 05 PM

PostAuthentication via Uas

Screenshot 2024-12-13 at 9 00 35 PM

Authorization Request to adyen with connector token received from uas

 
Screenshot 2024-12-13 at 9 01 20 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 1]
```diff
@@ -16,6 +16,7 @@ use error_stack::{report, ResultExt};
 use futures::FutureExt;
 #[cfg(all(any(feature = "v1", feature = "v2"), not(feature = "customer_v2")))]
 use hyperswitch_domain_models::payments::payment_intent::PaymentIntentUpdateFields;
+use hyperswitch_domain_models::router_request_types::unified_authentication_service::UasAuthenticationResponseData;
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm.rs:19]:
```suggestion
use hyperswitch_domain_models::router_request_types::unified_authentication_service;
```

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 2]
```diff
@@ -883,7 +886,7 @@ impl<F: Clone + Send + Sync> Domain<F, api::PaymentsRequest, PaymentData<F>> for
             business_profile,
         ))
         .await?;
-
+        println!("logg1 {:?}", payment_data.payment_method_data);
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm.rs:889]: remove println

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 3]
```diff
@@ -1102,19 +1105,84 @@ impl<F: Clone + Send + Sync> Domain<F, api::PaymentsRequest, PaymentData<F>> for
...
+                                card_network: None,
+                                card_type: None,
+                                card_issuing_country: None,
+                                bank_code: None,
+                                nick_name: None,
+                                eci: authentication_details.eci,
+                            },
+                        )
+                    }
+                    Ok(UasAuthenticationResponseData::PreAuthentication {}) => {
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1147]: For PreAuthentication we creating authentication entry?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 4]
```diff
@@ -1102,19 +1105,84 @@ impl<F: Clone + Send + Sync> Domain<F, api::PaymentsRequest, PaymentData<F>> for
                 )
                 .await?;
 
-                payment_data.payment_method_data =
-                    network_token.map(domain::PaymentMethodData::NetworkToken);
+                let network_token = match response.response.clone() {
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1108]: Return status from the branch and have one                         unified_authentication_service::create_new_authentication(

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Review by likhinbopanna — APPROVED** [review]

**Pull request merged**